Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\rightarrow$ Restart) and then **run all cells** (in the menubar, select Cell $\rightarrow$ Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [ ]:
NAME = "Akshata Padalkar"
COLLABORATORS = ""

---

Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\to$ Restart) and then **run all cells** (in the menubar, select Cell $\to$ Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [ ]:
NAME = "Akshata Padalkar"
COLLABORATORS = ""

---

<div class="alert alert-block alert-info">
<h3>Student Information</h3> Please provide information about yourself.<br>
<b>Name</b>: Akshata Padalkar<br> 
<b>NetID</b>: ANP178<br>
<b>Recitation #</b>:2<br>
<b>Notes to Grader</b> (optional):<br>
<br><br>
<b>IMPORTANT</b>
Your work will not be graded withour your initials below<br>
I certify that this lab represents my own work and I have read the RU academic intergrity policies at<br>
<a href="https://www.cs.rutgers.edu/academic-integrity/introduction">https://www.cs.rutgers.edu/academic-integrity/introduction </a><br>
<b>Initials</b>: AP     (eg: NC for Naina Chaturvedi)

# CS 439 - Introduction to Data Science
### Spring 2025

## Lab 1 - Pandas, NumPy (100 points)
### Due Date: Feb 19th, 2025

Welcome to your first lab in Introduction to Data Science! In this lab, you will explore two fundamental libraries in Python: NumPy and Pandas. This notebook has multiple tasks that you need to complete. To complete a task, just double click on the cell (notebook must be opened in Jupyter/CodeBench) and enter your answer. Be sure to save the notebook often (ctrl-S or cmd-S) and execute the cell (Shift-Enter) to make sure you code is saved and executes correctly.

### Lab Discussions
Post any questions regarding this lab to canvas discussion board on Lab 01 or visit office hours.

## Course Policies

**Collaboration Policy**

Data science is a collaborative activity. While you may talk with others about the labs, we ask that you **write your solutions individually**. You are not allowed to copy code from other notebooks. Attempting to submit code that is not yours can result in penalties and academic disciplinary actions. Please consult course staff for any clarification regarding collaboration policies.

**AI Policy**

We **only allow** the use of responsible **AI tool youchat** to help you find information as you write code to solve labs. The use of AI should be done cautiously. It **does not** free you from understanding the code you write. Being able to read, interpret and test code written by AI is required for you to answer questions in quizzes and exams. **DO NOT use** other AI tools such as chatGPT and Bard.

## Part 1: NumPy (40 Points)

In [40]:
import numpy as np

### Task 1: Inventory Data Reshaping (5 Points)

- Task: A warehouse stores its inventory in a 3x4 grid where each cell represents the quantity of a product. Simulate this inventory using an array with random integers between 10 and 50. Then, reshape the array into a 2x6 grid.

- Significance: Reshaping is crucial in numerical computations and machine learning to align data structures for efficient processing.

- Output: Display the reshaped array, and it's shape



In [41]:
# Code goes here
A = np.random.randint(10, 50, (3, 4))
#print("A =\n", A) 
reshaped_A = A.reshape(2, 6)
print("reshaped_A = \n", reshaped_A)
print("Shape: ", reshaped_A.shape)

reshaped_A = 
 [[46 43 11 35 45 36]
 [16 10 35 32 33 22]]
Shape:  (2, 6)


### Task 2: Structural Engineering Matrix (10 Points)

- Task: A structural engineer wants to model a 5x5 stiffness matrix where the main diagonal represents primary support (value 1), and the upper and lower subdiagonals represents secondary support (value 2). Construct this matrix.

- Significance: Diagonal matrices are widely used in engineering and physics to represent stiffness, resistance, or transformations.

- Output: Display the matrix, and it's shape

In [42]:
# Code goes here
A = np.zeros((5,5),int)                                               
A[range(5),range(5)]=1                                               
A[range(4),range(1,5)]=2                                              
A[range(1,5),range(4)]=2                                              
print(A)    
print("Shape: ", A.shape)

[[1 2 0 0 0]
 [2 1 2 0 0]
 [0 2 1 2 0]
 [0 0 2 1 2]
 [0 0 0 2 1]]
Shape:  (5, 5)


### Task 3: Data Combination for Analysis (10 Points)
- Task: A research study collects two separate 1D datasets of size 5. Combine these datasets into one and organize them into a 2x5 format. Also find the mean, median, and standard deviation of the combined dataset.

- Significance: Combining datasets is fundamental in data science, allowing integration of multiple sources of information.

- Output: Display the combined array, it's shape and the it's mean, median, and standard deviation.

In [43]:
data1 = np.array([12.4, 4.1, 45.2, 4.3, 44.1])
data2 = np.array([1, 45, 3.5, 66.2, 3])

# Code goes here
matrix = np.array([data1, data2])

print("Matrix: \n", matrix)
print()
mean = np.mean(matrix)
median = np.median(matrix)
std = np.std(matrix)
print("Mean: ", mean,"\nMedian: ", median,"\nStandard Deviation: ", std)
print("Shape: ", matrix.shape)

Matrix: 
 [[12.4  4.1 45.2  4.3 44.1]
 [ 1.  45.   3.5 66.2  3. ]]

Mean:  22.880000000000003 
Median:  8.35 
Standard Deviation:  23.176833260823187
Shape:  (2, 5)


### Task 4: Linear Transformations in Computer Graphics (5 Points)

- Task: Compute the transformed position when matrix A is applied to vector X, followed by translation b.

- Significance: Linear transformations are essential for scaling, rotating, and translating objects in space.

- Hint: Find `Ax + b` using Matrix multiplication and broadcasting

- Output: Display the transformed matrix


In [44]:
A = np.array([[2, 3], [4, 5]])
x = np.array([1, 2])
b = np.array([5, 6])

# Code goes here
dot = np.dot(A,x)
add = np.add(dot,b)
print(add)

[13 20]


### Task 5: Signal Processing Correlation (5 Points)

- Task: Compute the dot product correlation between two 3x3 matrices to measure similarity, commonly used in image processing and machine learning.

- Significance: The dot product quantifies how similar two matrices are, crucial in pattern recognition.

- Output: Display the dot product of the 2 matrices




In [45]:
filter1 = np.array([
    [1,1,1],
    [1,0,1],
    [1,1,1]
])

filter2 = np.array([
    [1,1,2],
    [1,2,1],
    [2,1,1]
])

# Code goes here
dot = np.sum(filter1*filter2) 
print(dot)

10


### Task 6: Solving Electrical Circuit Equations (15 Points)

- Task: Solve the linear equation system $V = IR$ to determine Current($I$) in the circuit, given the Voltage($V$) and Resistor($R$) values

- Significance: Solving linear systems is key in engineering, physics, and optimization models.

- Hint: Use [`np.linalg` library](https://numpy.org/doc/2.1/reference/routines.linalg.html) to solve the system of equations $R$ and $V$ to find $I$

- Output: Display the Current values


In [46]:
V = np.array([[4,4,0]])
R = np.array([
    [13, -12, 3],
    [1, 4, 1],
    [3, -4, -1]
])

new_V = V.reshape(-1, 1)
# Code goes here
I = np.linalg.inv(R) @ new_V
print(I)

[[1.  ]
 [0.75]
 [0.  ]]


## Part 2: Pandas (60 Points)

In this section you **MUST** reuse the results from previous question wherever possible. **DO NOT RE-WRITE CODE**

In [47]:
import pandas as pd

### Task 1: Reading CSV files (5 Points)

- Task: Load and display the first five rows of the StudentPerformance.csv data to inspect its structure.

- Significance: Reading data is the first step in analysis, allowing us to explore data organization.

In [48]:
# Code goes here
df = pd.read_csv('StudentsPerformance.csv')
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A                         NaN  free/reduced   
4    male        group C                some college           NaN   

  test preparation course  math score  reading score  writing score  
0                    none         NaN           72.0           74.0  
1               completed        69.0           90.0           88.0  
2                    none        90.0            NaN           93.0  
3                     NaN         NaN           57.0           44.0  
4                     NaN        76.0            NaN            NaN

### Task 2: DataFrame Inspection (10 Points)

- Task: Display key details such as data types, missing values, and overall structure. (5 Points)

- Significance: Understanding data structure ensures appropriate cleaning and transformations.


In [49]:
# Code goes here
print("Data Types:\n", df.dtypes)
print("\nMissing Values:\n", df.isnull().sum())
print("\nSummary Statistics:\n", df.describe())

Data Types:
 gender                          object
race/ethnicity                  object
parental level of education     object
lunch                           object
test preparation course         object
math score                     float64
reading score                  float64
writing score                  float64
dtype: object

Missing Values:
 gender                         101
race/ethnicity                  96
parental level of education     95
lunch                          100
test preparation course         85
math score                     106
reading score                   98
writing score                  115
dtype: int64

Summary Statistics:
        math score  reading score  writing score
count  894.000000     902.000000     885.000000
mean    66.360179      69.205100      68.058757
std     15.278820      14.484676      15.126351
min      0.000000      17.000000      15.000000
25%     57.000000      59.000000      57.000000
50%     67.000000      70.000000      69

What did you observe in the data? (5 Points) </br></br>
Several columns have missing values, particularly writing score (115 missing), math score (106 missing), and gender (101 missing). The missing values could impact analysis and may require imputation or removal of incomplete rows. The presence of 0 in math scores might indicate missing data encoded as zeros or students who did not attempt the test, which can be an outlier. Categorical columns such as gender, race/ethnicity, parental level of education, lunch, and test preparation course are stored as objects. The math score, reading score, and writing score columns are numerical (float64), which is expected.

### Task 3: Data Cleaning (10 Points)

- Task: Remove incomplete rows and reset the index (if needed) to ensure data integrity.

- Significance: Cleaning data improves accuracy in analysis and machine learning models.

In [50]:
# Code goes here
print(len(df))
#assumed data did not need to be preserved or imputed. 
df_filtered = df.dropna().reset_index(drop=True)
df_filtered.to_csv('final_clean.csv', index=False)

print(len(df_filtered))
print(df_filtered.isnull().sum())
df_filtered.head()

1000
440
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64


gender race/ethnicity parental level of education         lunch  \
0  female        group C                some college      standard   
1  female        group B                some college      standard   
2    male        group B                some college  free/reduced   
3    male        group D                 high school  free/reduced   
4    male        group D          associate's degree      standard   

  test preparation course  math score  reading score  writing score  
0               completed        69.0           90.0           88.0  
1               completed        88.0           95.0           92.0  
2                    none        40.0           43.0           39.0  
3               completed        64.0           64.0           67.0  
4                    none        40.0           52.0           43.0

### Task 4: Filtering Data (10 Points)

- Task: Identify students who scored above 80 in math and completed the test preparation course.

- Significance: Filtering helps extract relevant subsets of data for focused analysis.

- Output: Display all details of the filtered students

In [51]:
# Code goes here
#Assuming we are to use the cleaned dataframe to complete this as stated at the beginning of this portion of tasks "you MUST reuse the results from previous question wherever possible"

high_math_scores = df_filtered[(df_filtered["math score"] > 80) & (df_filtered["test preparation course"] == "completed")]
print(high_math_scores)

     gender race/ethnicity parental level of education         lunch  \
1    female        group B                some college      standard   
16     male        group E          associate's degree      standard   
24     male        group C                 high school      standard   
43     male        group C                some college      standard   
49     male        group B          associate's degree      standard   
50   female        group C                some college      standard   
54     male        group E                some college  free/reduced   
59     male        group E          associate's degree  free/reduced   
67     male        group B          associate's degree  free/reduced   
70   female        group C           bachelor's degree      standard   
75   female        group C             master's degree      standard   
77   female        group D            some high school      standard   
94   female        group E          associate's degree  free/red

### Task 5: Using `apply` Function (10 Points)

- Task: Categorize students based on math scores into "High", "Medium", and "Low" categories (Add the details to the original Data Frame)

- Significance: The apply function is useful for transforming and categorizing data efficiently.

- Output: Display the top 10 rows of the data after modifying the Data Frame


In [52]:
# Code goes here
#Due to mixed signal directions where it is stated"you MUST reuse the results from previous question wherever possible" and Add the details to the original Data Frame, I have decided to add it to both dataframes and display both as well.  
def math_category(math):
    if math <= 25:
        return 'Low'
    elif math >= 76:
        return 'High'
    else:
        return 'Medium'

df_filtered['math_category'] = df_filtered['math score'].apply(math_category)#cleaned dataframe
df['math_category'] = df['math score'].apply(math_category)#original datafram
df.head(10)#original
#print(df_filtered.head(10))#cleaned

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A                         NaN  free/reduced   
4    male        group C                some college           NaN   
5  female        group B                         NaN      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course  math score  reading score  writing score  \
0                    none         NaN           72.0           74.0   
1               completed        69.0           90.0           88.0   
2                    none        90.0            NaN           93.0   
3                     NaN         NaN           57.0           44.0   
4                     NaN        76.0            NaN            NaN   
5                    none        71.0           83.0            NaN   
6               completed        88.0           95.0           92.0   
7                    none        40.0           43.0           39.0   
8               completed        64.0           64.0           67.0   
9                    none         NaN           60.0           50.0   

  math_category  
0        Medium  
1        Medium  
2          High  
3        Medium  
4          High  
5        Medium  
6          High  
7        Medium  
8        Medium  
9        Medium

### Task 6: GroupBy Analysis (15 Points)

- As a Data Scientist, on of your core responsibilites is to extract meaningful insights by asking the right questions from the data and finding answers to them. The `groupby` operation in Pandas enables you to explore and find answers to various analytical questions efficiently

- For each of the following questions, extract answers from the data

1. How are students performing across different parental education levels? (use mean) - 5 Points
2. What is the average test score for each gender? - 5 Points
3. Which test has the highest average scores? - 5 Points

In [53]:
# Code for Question 1 
df_filtered.groupby('parental level of education')[['math score','reading score','writing score']].mean()

math score  reading score  writing score
parental level of education                                          
associate's degree            67.597938      71.422680      71.195876
bachelor's degree             69.153846      73.365385      73.538462
high school                   62.350000      63.862500      61.825000
master's degree               70.193548      76.225806      76.258065
some college                  68.340000      70.550000      69.640000
some high school              62.037500      67.487500      65.675000

In [54]:
# Code for Question 2
df_filtered.groupby('gender')[['math score','reading score','writing score']].mean()

math score  reading score  writing score
gender                                          
female   64.318966      73.728448      73.577586
male     68.230769      65.211538      63.403846

In [55]:
# Code for Question 3
avg_scores = df_filtered[['math score', 'reading score', 'writing score']].mean()
highest_avg_test = avg_scores.idxmax()
highest_avg_value = avg_scores.max()
print(avg_scores)
print("The test with the highest average score is:", highest_avg_test, highest_avg_value)

math score       66.168182
reading score    69.702273
writing score    68.768182
dtype: float64
The test with the highest average score is: reading score 69.70227272727273


What are a few other questions you can ask from the data? </br></br>
Some questions are how does the test preparation course impact scores? Which race/ethnicity group scores the highest on average? Does lunch type affect scores?

### Task 7: Concatenation (10 Points)

- Task: Extract and combine the top and bottom 10 students by concatenating the subsets.

- Significance: Concatenation helps in segmenting and appending datasets for comparison.

- Output: Display the concatenated set

In [56]:
# Code goes here
#Assuming we are to use the cleaned dataframe to complete this as stated at the beginning of this portion of tasks "you MUST reuse the results from previous question wherever possible"

average_cat = df_filtered['average score'] = df_filtered[['math score', 'reading score', 'writing score']].mean(axis=1)

sorted_df = df_filtered.sort_values(by='average score', ascending=False)#(top 10)
sort_df = df_filtered.sort_values(by='average score', ascending= True)#(bottom 10)
top_10_students = sorted_df.head(10)
bottom_10_students = sort_df.head(10)

combined_students = pd.concat([top_10_students, bottom_10_students], keys=['Top 10', 'Bottom 10'])

print(combined_students)


               gender race/ethnicity parental level of education  \
Top 10    210  female        group E           bachelor's degree   
          77   female        group D            some high school   
          70   female        group C           bachelor's degree   
          59     male        group E          associate's degree   
          418  female        group D             master's degree   
          251  female        group E           bachelor's degree   
          312  female        group C          associate's degree   
          260  female        group C           bachelor's degree   
          205  female        group E                some college   
          243  female        group A            some high school   
Bottom 10 261    male        group B                 high school   
          8    female        group B            some high school   
          264  female        group C                 high school   
          345  female        group B            

## Feedback
What do you expect to learn from this course? </br></br>
I am excited to further develop my skills in python and get a great and useful introduction to deep learning and LLMs.

<div class="alert alert-block alert-info">
<h2>Submission Instructions</h2> 
<b> File Name:</b> Please rename the file as netID_Lab01_Working with Numpy and Pandas.ipynb (eg: netID = adg133)<br>
<b> Submit through: </b> CodeBench <br>
<b>Warning:</b> Failure to follow directions may result in loss points.<br>
</div>